librad/src/git/storage.rs

```diff
@@ -108,6 +109,24 @@ impl AsRef<git2::Repository> for Storage {
     }
 }
 
+pub struct LinearHistoryCommits<'a>(Option<git2::Commit<'a>>);
+
+impl<'a> Iterator for LinearHistoryCommits<'a> {
```

Comment:
Now that I have seen `Revwalk` I see that I could implement this using a revwalk at its core.

However I explicitly want to ensure that each commit has no more than one parent, and doing it with a revwalk means filtering each step...at that point `LinearHistoryCommits` with its 10 lines is not that different.

Reply:
Just to eliminate ambiguity: what if the 1st parent points to a bogus document and the second one to the correctly hashed one?
We could resolve the situation but the code needs to be written and there's no reason to create that situation in the first place.
Why would anyone put a commit with more than one parent in an entity history?

Reply:
Now that I think of it, for "items" (other than entities) we'd want them to refer to their "dependencies", and the reference should also be expressed as a parent commit so that the dependencies are pulled and preserved even if the original repo disappears (zero mirrors online, at all).

But for entities, we are already preserving them all in the monorepo, so additional parents are not needed, they would not even be possible because outside of the monorepo they would need to point to a different repo...

Reply:
> what if the 1st parent points to a bogus document and the second one to the correctly hashed one?

The order of the parents is not arbitrary -- the first one is the head of the branch _into_ which one merges.

> Why would anyone put a commit with more than one parent in an entity history?

Assume the flow for collecting the signatures on a new revision is similar to a PR: everyone publishes their signed version, and then they merge all the single-signature ones into one. That could very well be a merge-commit, for various reasons (documentation, for example) -- the point is that the merge result has to yield a tree, which yields an "id" file, which is valid. This is all we're interested in.

> additional parents [...] would not even be possible

That's not true, additional parents will be forced to be in the object database of a standalone export -- which is kind of the point, standalone repos should be self-contained, the monorepo is for efficiency.

Reply:
Ok, I get it, this is really similar to the "items" case in the end.

Could we mandate that the "significant" parent is **always** the first, and the others are there only to speed up key dissemination and for "documenting merges"?
Which means that the verification algorithm will use the 1st commit if present and ignore the rest.

Reply:
I think the `revwalk` will always pick the significant parent, i.e. it's direct ascendant. At least that's the behaviour I saw when I was implementing the commit history feature in surf.

Reply:
Yes, what I mean is to be explicit about this in the specification how entities are stored.

A commit could have multiple parents but only the 1st is "significant" (it points to the previous revision).
And we should also specify what the other parents _could_ point to.

I might be paranoid but adding a parent referring to an entirely unrelated entity should have no effect but could have unintended indirect effects, in extreme cases wasting storage or network resources (a kind of mild DoS attack).
I'd have to wrap my head around it to understand all the consequences.
I was so strict about parents also because I was trying to avoid analyzing this in full :-)

Reply:
My point is: we are looking at the _blob data_, not at the commits. Thus, the topology of the history graph does not actually matter, as long as we can follow a tip backwards along _one_ path, until there are no more parents. If invalid blobs occur in this history, the verification should just fail. 

To illustrate what I mean:

Suppose we have two branches:

    refs/heads/A -> c1
    refs/heads/B -> c2

We want to make `c2` reachable from `refs/heads/A`, so we say we "merge B into A". Slightly simplified, the following cases can occur:

1. `c1 == c2`

    Nothing happens, both commits are equal

2. `[c1] == parents(c2)`

    Trivial: `c2` is just a commit on top of `c1`. `git` simply updates `refs/heads/A -> c2`.

3. `c1 ∈ parents(c2)`

    This is a fast-forward, we only need to update the branch ref. It is indeed possible that the first parent is not equal to `c1` (I was wrong earlier) -- typically because we merged `A` into `B` previously. In this case, `refs/heads/A` will point to a merge commit (ie. a commit with more than one parent). Unless that merge commit contains an invalid state (ie. you committed the merge without resolving conflicts), **it does not matter** -- examining the `tree` will yield valid data.

4. `parents(c2).any(|parent| parent.is_reachable_from(c1))`

    What we usually call a "merge": `B` at some point built on top of `A`, but `A` received other commits which are not reachable from `B`. Here, we will get a merge commit where the first parent is `c1`, and the second one is `c2`.

    It is potentially confusing at this point that `git log` by default shows the history in chronological order (more specifically: "Show no parents before all of its children are shown, but otherwise show commits in the commit timestamp order") -- `c2` may appear near the point `B` branched off of `A`, while the first commit shown is the merge commit. Again, **it does not matter**: if we examine the tree of the first commit, what we get is the _combined_ result. If that is not valid according to our rules, the merge was bad.

  This is also why we should use `simplify_first_parent` (ie. `git log --first-parent`): there is no guessing of unreliable timestamps, the path leads through all the merges, potentially omitting commits which may otherwise yield conflicts.

Of course, this does not mean that one could use `git merge` directly (without some serious trickery, that is) -- due to the internal `previous_hash` and (redundant) `revision`, non-fast-forwards will **always** produce a conflict acc. to our rules. So effectively, the merge commit's `tree` would contain a newly signed version of the document anyways (but `--first-parent` ignores our first attempt, so that's fine).

So what's the point, you ask? Simple: we reserve the `parents` field for future use, in any way we see fit.

But ok, you say, this all seems a bit complicated with all this hash-linking and caching and stuff, why can't we just use a single-parent git history? Yeah, I thought the same. The idea is that each signer signs only the _data payload_, not the signatures of other signers (or other metadata). That's not possible with normal GPG signing of commits, because a. there can only be one signature, and b. what is signed is the entire commit object (minus the signature). Typically, the way this is done in git without interfering with any assumptions the tooling might have, is through [trailers](https://git-scm.com/docs/git-interpret-trailers) (you know, `Signed-off-by` et. al.). I am starting to think that this may yield an equivalent, but way simpler system.